src/main/java/io/nats/client/impl/NatsJetStreamManagement.java

```diff
@@ -414,18 +408,21 @@ public void _requestMessageBatch(String streamName, MessageBatchGetRequest messa
 
                 MessageInfo messageInfo = new MessageInfo(msg, streamName, true);
                 handler.onMessageInfo(messageInfo);
-                timeLeft = maxTimeMillis - (System.currentTimeMillis() - start);
             }
-        } catch (InterruptedException e) {
+        }
+        catch (InterruptedException e) {
             // sub.nextMessage was fetching one message
             // and data is not completely read
             // so it seems like this is an error condition
             Thread.currentThread().interrupt();
             throw new RuntimeException(e);
         } finally {
-            try {
-                handler.onMessageInfo(MessageInfo.EOD);
-            } catch (Exception ignore) {
+            if (sendEod) {
+                try {
+                    handler.onMessageInfo(MessageInfo.EOD);
+                }
+                catch (Exception ignore) {
+                }
```

Comment:
don't send the EOD if the request didn't ask (fetch) or in the case of an abnormal status